File: 0494-target-sum/0494-target-sum.cpp
class Solution {
public:
    int N, dp[21][3001];
    int dfs(vector<int>& nums, int ind, int s) {
        if(ind==N) {
            if(s == 0) return 1;
            else return 0;
        }
        if(dp[ind][s+1000]!=-1) return dp[ind][s+1000];
        int res = 0;
        res = dfs(nums, ind+1, s-nums[ind]) + dfs(nums, ind+1, s+nums[ind]);
        return dp[ind][s+1000] = res;
    }
    int findTargetSumWays(vector<int>& nums, int s) {
        N = nums.size();
        if(s>1000) return 0;
        memset(dp,-1,sizeof(dp));
        return dfs(nums, 0, s);
    }
};

File: 0494-target-sum/0494-target-sum.java
class Solution {
    public int findTargetSumWays(int[] nums, int target) {
        int n = nums.length;
        // count the number of partitions such that the difference between them is target
        int sum = 0;
        for(int i:nums) sum+=i;
        if(sum<target) return 0;
        /**
            divide into 2 partitions S1 and S2
            where sum of S1 and S2 :
            s2 + s1 = total sum
            s2 = totalSum - s1    - (i)
            and since s1>s2:
            s2 = s1 - target      -(ii)
            add eqn. (i) and (ii):
            2*s2 = sum - target
        */
        if((sum-target)%2==1) return 0;
        sum = (sum-target)/2;
        int[][] dp = new int[n+1][sum+1];
        dp[0][0]=1;
        for(int i=1;i<=n;i++) {
            for(int j=0;j<=sum;j++) {
                int dont=dp[i-1][j];
                int take=0;
                if(j-nums[i-1]>=0) {
                    take=dp[i-1][j-nums[i-1]];
                }
                dp[i][j]=take+dont;
            }
        }
        return dp[n][sum];
    }
}